spock-core/src/main/java/org/spockframework/mock/IResponseGenerator.java

```diff
@@ -16,10 +16,12 @@
 
 import org.spockframework.util.Beta;
 
+import java.util.function.Supplier;
+
 /**
  * Generates responses to mock invocations.
  */
 @Beta
 public interface IResponseGenerator {
-  Object respond(IMockInvocation invocation);
+  Supplier<Object> respond(IMockInvocation invocation);
```

Comment:
Shouldn't we add the supplier as default method and keep the old one?

```suggestion
  @Nullable
  default Object respond(IMockInvocation invocation) {
    return null;
  }
  
  /**
   * This method will be called by consumers.
   */
  default Supplier<Object> responseSupplier(IMockInvocation invocation) {
    return () -> respond(invocation); 
  }
```

Reply:
I'm not so sure how it is best.

We have basically 2 cases in 3rd party code.
1. someone is calling `respond`.
2. someone is implementing `respond`.

For code implementing respond, it would probably just work with your suggested change, as we use the `responseSupplier` which will call their `respond` method.

For code calling `respond`, for example like in `DetachedMockFactoryDocSpec` or `StaticSpyDocSpec`, with your suggested change it could get the wrong value as the `responseSupplier` is not used. Without your change already compiled code would fail as the signature changed. The code would actually still compile if just re-compiled but hopefully tests would recognize that behavior changed.

So maybe it is indeed beste to change it like you suggested and in those that really need the supplier to be used throw an exception in `respond`? This way existing code implementing `respond` would continue to work and code calling respond would work as long as they only call methods where it does not make a difference, but if they call one of the three where it makes a difference would get an exception and need to change their code.

What do you think?

Reply:
In case you started to look at the latest changes already, please do it again, sorry.
After having slept it over, I was unhappy with what I did yesterday and reworked it again.